### PR TITLE
Dynamically add robots.txt and favicon.ico per root page

### PR DESCRIPTION
**Pull request by Toflar**
This PR provides two new routes:

* `/favicon.ico` which provides the icon you can now select in the root page settings (the fallback one).
* `/robots.txt` which provides the `robots.txt` for any root page (the fallback one). Users can manually enter content plus we have an event that allows us to extend it dynamically. The core uses it to dynamically add all `sitemap.xml` entries making it super convenient for users 🎉

### CODE REVIEW

File: core-bundle/src/Resources/config/listener.yml [hunk 1]
```diff
@@ -16,6 +16,13 @@ services:
         tags:
             - { name: kernel.event_listener, event: kernel.terminate, method: onKernelTerminate }
 
+    contao.listener.robots_txt:
```
Comment (leofeyer) [inline core-bundle/src/Resources/config/listener.yml]: We have agreed on using the class name as service key for all new services, so this needs to become `Contao\CoreBundle\EventListener\RobotsTxtListener`.

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

File: core-bundle/src/Controller/FaviconController.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    private $responseTagger;
+
+    public function __construct(ContaoFramework $contaoFramework, ResponseTagger $responseTagger = null)
+    {
+        $this->contaoFramework = $contaoFramework;
+        $this->responseTagger = $responseTagger;
+    }
+
+    /**
+     * @Route("/favicon.ico", name="contao_favicon")
```
Comment (aschempp) [inline core-bundle/src/Controller/FaviconController.php:46]: does the route really need a route name? I mean noone will ever generate that?

File: core-bundle/src/Event/ContaoCoreEvents.php [hunk 3]
```diff
@@ -68,6 +68,15 @@ final class ContaoCoreEvents
...
     public const PREVIEW_URL_CONVERT = 'contao.preview_url_convert';
 
+    /**
+     * The contao.robots_txt event is triggered when the /robots.txt route is called.
+     *
+     * @var string
+     *
+     * @see RobotsTxtEvent
+     */
+    public const ROBOTS_TXT = 'contao.robots_txt';
```
Comment (aschempp) [inline core-bundle/src/Event/ContaoCoreEvents.php:78]: It is deprecated in Symfony to use event names. The event class should be the name…

Comment (Toflar) [reply core-bundle/src/Event/ContaoCoreEvents.php:78]: Yes but I cannot just change that atm as we do not require 4.4 yet.

File: core-bundle/src/Resources/config/services.yml [hunk 4]
```diff
@@ -28,6 +28,22 @@ services:
             - { name: container.service_subscriber }
         public: true
 
+    Contao\CoreBundle\Controller\FaviconController:
+        arguments:
+            - '@contao.framework'
+            - '@?fos_http_cache.http.symfony_response_tagger'
+        tags:
+            - controller.service_arguments
```
Comment (aschempp) [inline core-bundle/src/Resources/config/services.yml:36]: The `controller.service_arguments` tag is not needed if you don't get services in your action methods.

File: core-bundle/src/Controller/FaviconController.php (hunk 2 above)
Comment (Toflar) [reply core-bundle/src/Controller/FaviconController.php:46]: Imho a route should have a name yes.

File: core-bundle/src/Event/ContaoCoreEvents.php (hunk 3 above)
Comment (aschempp) [reply core-bundle/src/Event/ContaoCoreEvents.php:78]: why not? Just use `event::class` as the event name?

File: core-bundle/src/Controller/FaviconController.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Controller/FaviconController.php:46]: https://symfony.com/doc/current/routing.html#generating-urls

> If you don't set the route name explicitly with the name option, Symfony generates an automatic name based on the controller and action.

I think that would be totally fine in this case, noone ever will override the named route, as it is not used to generate a route anyway.

File: core-bundle/src/Resources/config/services.yml (hunk 4 above)
Comment (fritzmg) [reply core-bundle/src/Resources/config/services.yml:36]: afaik it is still needed for dependency injection in the constructor.

Comment (aschempp) [reply core-bundle/src/Resources/config/services.yml:36]: No, what you mean is the service locator. But this controller does not extend from `AbstractController` so service locator is not used at all.

Comment (fritzmg) [reply core-bundle/src/Resources/config/services.yml:36]: My custom controllers do not extend from `AbstractController` either and without the `controller.service_arguments`, no services would be injected into the constructor of the controller.

Comment (fritzmg) [reply core-bundle/src/Resources/config/services.yml:36]: Hm, or not, just checked it again. Setting the controller to `public: true` is enough.

Comment (Toflar) [reply core-bundle/src/Resources/config/services.yml:36]: Guys, this PR is merged. If you want changes, create a new PR. Nothing will happen here.

Comment (leofeyer) [reply core-bundle/src/Resources/config/services.yml:36]: I'm still following the discussion though.

Comment (leofeyer) [reply core-bundle/src/Resources/config/services.yml:36]: The `__invoke()` method needs the `Request` object. Will this still work without the tag?

**Pull request merged**